server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,747 @@
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetBytesRead();
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+            // It's expensive to look up every term and visit every document of the postings lists of all terms.
+            // As we track the min/max positions of read bytes, we just visit the two ends of a partition containing
+            // the data. We might miss some small parts of the data, but it's an good trade-off to speed up the process.
+            TermsEnum termsEnum = terms.iterator();
+            if (terms instanceof FieldReader) {
+                final BlockTermState minState = Objects.requireNonNull(
+                    getBlockTermState(termsEnum, terms.getMin()), "can't retrieve the block term state of the min term");
+                final BlockTermState maxState = Objects.requireNonNull(
+                    getBlockTermState(termsEnum, terms.getMax()), "can't retrieve the block term state of the max term");
+                final long skippedBytes = maxState.distance(minState);
+                stats.addInvertedIndex(field.name, skippedBytes);
+                termsEnum.seekExact(terms.getMax());
+                postings = termsEnum.postings(postings, PostingsEnum.ALL);
+                if (postings.advance(termsEnum.docFreq() - 1) != DocIdSetIterator.NO_MORE_DOCS) {
+                    readProximity(terms, postings);
+                }
+                final long bytesRead = directory.getBytesRead();
+                int visitedTerms = 0;
+                final long totalTerms = terms.size();
+                termsEnum = terms.iterator();
+                // Iterate until we really access the first terms, but iterate all if the number of terms is small
+                while (termsEnum.next() != null) {
+                    cancellationChecker.logEvent();
+                    ++visitedTerms;
+                    if (totalTerms > 1000 && visitedTerms % 50 == 0 && directory.getBytesRead() > bytesRead) {
+                        break;
+                    }
+                }
+            } else {
+                // slow mode: traverse every postings of all terms
+                while (termsEnum.next() != null) {
+                    cancellationChecker.logEvent();
+                    termsEnum.docFreq();
+                    termsEnum.totalTermFreq();
+                    postings = termsEnum.postings(postings, PostingsEnum.ALL);
+                    while (postings.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        readProximity(terms, postings);
+                    }
+                }
```

Comment:
why does the optimization of only reading postings and proximity for the first and last terms not work in the general case?

Reply:
I added a comment why we execute a slow mode for other posting implementations rather than block tree based impl.